// @parcel/core/lib/AssetGraphBuilder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _events = _interopRequireDefault(require("events"));

var _nullthrows = _interopRequireDefault(require("nullthrows"));

var _path = _interopRequireDefault(require("path"));

var _utils = require("@parcel/utils");

var _AssetGraph = _interopRequireDefault(require("./AssetGraph"));

var _RequestTracker = _interopRequireWildcard(require("./RequestTracker"));

var _constants = require("./constants");

var _ParcelConfig = _interopRequireDefault(require("./ParcelConfig"));

var _ParcelConfigRequest = _interopRequireDefault(require("./requests/ParcelConfigRequest"));

var _EntryRequest = _interopRequireDefault(require("./requests/EntryRequest"));

var _TargetRequest = _interopRequireDefault(require("./requests/TargetRequest"));

var _AssetRequest = _interopRequireDefault(require("./requests/AssetRequest"));

var _PathRequest = _interopRequireDefault(require("./requests/PathRequest"));

var _Validation = _interopRequireDefault(require("./Validation"));

var _ReporterRunner = require("./ReporterRunner");

var _dumpGraphToGraphViz = _interopRequireDefault(require("./dumpGraphToGraphViz"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const typesWithRequests = new Set(['entry_specifier', 'entry_file', 'dependency', 'asset_group']);

class AssetGraphBuilder extends _events.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "assetGraph", void 0);

    _defineProperty(this, "requestGraph", void 0);

    _defineProperty(this, "requestTracker", void 0);

    _defineProperty(this, "assetRequests", void 0);

    _defineProperty(this, "runValidate", void 0);

    _defineProperty(this, "queue", void 0);

    _defineProperty(this, "changedAssets", new Map());

    _defineProperty(this, "options", void 0);

    _defineProperty(this, "optionsRef", void 0);

    _defineProperty(this, "workerFarm", void 0);

    _defineProperty(this, "cacheKey", void 0);

    _defineProperty(this, "entries", void 0);

    _defineProperty(this, "initialAssetGroups", void 0);

    _defineProperty(this, "handle", void 0);
  }

  async init({
    options,
    optionsRef,
    entries,
    name,
    assetGroups,
    workerFarm
  }) {
    this.options = options;
    this.optionsRef = optionsRef;
    this.entries = entries;
    this.initialAssetGroups = assetGroups;
    this.workerFarm = workerFarm;
    this.assetRequests = []; // TODO: changing these should not throw away the entire graph.
    // We just need to re-run target resolution.

    let {
      hot,
      publicUrl,
      distDir,
      minify,
      scopeHoist
    } = options;
    this.cacheKey = (0, _utils.md5FromObject)({
      parcelVersion: _constants.PARCEL_VERSION,
      name,
      options: {
        hot,
        publicUrl,
        distDir,
        minify,
        scopeHoist
      },
      entries
    });
    this.queue = new _utils.PromiseQueue();
    this.runValidate = workerFarm.createHandle('runValidate');
    let changes = await this.readFromCache();

    if (!changes) {
      this.assetGraph = new _AssetGraph.default();
      this.requestGraph = new _RequestTracker.RequestGraph();
    }

    this.assetGraph.initOptions({
      onNodeRemoved: node => this.handleNodeRemovedFromAssetGraph(node)
    });
    this.requestTracker = new _RequestTracker.default({
      graph: this.requestGraph,
      farm: workerFarm,
      options: this.options
    });

    if (changes) {
      this.requestGraph.invalidateUnpredictableNodes();
      this.requestTracker.respondToFSEvents(changes);
    } else {
      this.assetGraph.initialize({
        entries,
        assetGroups
      });
    }
  }

  async build(signal) {
    this.requestTracker.setSignal(signal);
    let errors = [];
    let root = this.assetGraph.getRootNode();

    if (!root) {
      throw new Error('A root node is required to traverse');
    }

    let visited = new Set([root.id]);

    const visit = node => {
      if (errors.length > 0) {
        return;
      }

      if (this.shouldSkipRequest(node)) {
        visitChildren(node);
      } else {
        this.queueCorrespondingRequest(node).then(() => visitChildren(node), error => errors.push(error));
      }
    };

    const visitChildren = node => {
      for (let child of this.assetGraph.getNodesConnectedFrom(node)) {
        if ((!visited.has(child.id) || child.hasDeferred) && this.assetGraph.shouldVisitChild(node, child)) {
          visited.add(child.id);
          visit(child);
        }
      }
    };

    visit(root);
    await this.queue.run();

    if (errors.length) {
      throw errors[0]; // TODO: eventually support multiple errors since requests could reject in parallel
    }

    (0, _dumpGraphToGraphViz.default)(this.assetGraph, 'AssetGraph'); // $FlowFixMe Added in Flow 0.121.0 upgrade in #4381

    (0, _dumpGraphToGraphViz.default)(this.requestGraph, 'RequestGraph');
    let changedAssets = this.changedAssets;
    this.changedAssets = new Map();
    return {
      assetGraph: this.assetGraph,
      changedAssets: changedAssets
    };
  } // TODO: turn validation into a request


  async validate() {
    let {
      config: processedConfig,
      cachePath
    } = (0, _nullthrows.default)((await this.requestTracker.runRequest((0, _ParcelConfigRequest.default)())));
    let config = new _ParcelConfig.default(processedConfig, this.options.packageManager, this.options.autoinstall);
    let trackedRequestsDesc = this.assetRequests.filter(request => {
      return config.getValidatorNames(request.filePath).length > 0;
    }); // Schedule validations on workers for all plugins that implement the one-asset-at-a-time "validate" method.

    let promises = trackedRequestsDesc.map(request => this.runValidate({
      requests: [request],
      optionsRef: this.optionsRef,
      configCachePath: cachePath
    })); // Skip sending validation requests if no validators were configured

    if (trackedRequestsDesc.length === 0) {
      return;
    } // Schedule validations on the main thread for all validation plugins that implement "validateAll".


    promises.push(new _Validation.default({
      requests: trackedRequestsDesc,
      options: this.options,
      config,
      report: _ReporterRunner.report,
      dedicatedThread: true
    }).run());
    this.assetRequests = [];
    await Promise.all(promises);
  }

  shouldSkipRequest(node) {
    return node.complete === true || !typesWithRequests.has(node.type) || node.correspondingRequest != null && this.requestGraph.getNode(node.correspondingRequest) != null && this.requestTracker.hasValidResult(node.correspondingRequest);
  }

  queueCorrespondingRequest(node) {
    switch (node.type) {
      case 'entry_specifier':
        return this.queue.add(() => this.runEntryRequest(node.value));

      case 'entry_file':
        return this.queue.add(() => this.runTargetRequest(node.value));

      case 'dependency':
        return this.queue.add(() => this.runPathRequest(node.value));

      case 'asset_group':
        return this.queue.add(() => this.runAssetRequest(node.value));

      default:
        throw new Error(`Can not queue corresponding request of node with type ${node.type}`);
    }
  }

  async runEntryRequest(input) {
    let request = (0, _EntryRequest.default)(input);
    let result = await this.requestTracker.runRequest(request);
    this.assetGraph.resolveEntry(request.input, result.entries, request.id);
  }

  async runTargetRequest(input) {
    let request = (0, _TargetRequest.default)(input);
    let result = await this.requestTracker.runRequest(request);
    this.assetGraph.resolveTargets(request.input, result.targets, request.id);
  }

  async runPathRequest(input) {
    let request = (0, _PathRequest.default)(input);
    let result = await this.requestTracker.runRequest(request);
    this.assetGraph.resolveDependency(input, result, request.id);
  }

  async runAssetRequest(input) {
    this.assetRequests.push(input);
    let request = (0, _AssetRequest.default)({ ...input,
      optionsRef: this.optionsRef
    });
    let assets = await this.requestTracker.runRequest(request);

    if (assets != null) {
      for (let asset of assets) {
        this.changedAssets.set(asset.id, asset);
      }

      this.assetGraph.resolveAssetGroup(input, assets, request.id);
    }
  }

  handleNodeRemovedFromAssetGraph(node) {
    if (node.correspondingRequest != null) {
      this.requestTracker.removeRequest(node.correspondingRequest);
    }
  }

  respondToFSEvents(events) {
    return this.requestGraph.respondToFSEvents(events);
  }

  getWatcherOptions() {
    let vcsDirs = ['.git', '.hg'].map(dir => _path.default.join(this.options.projectRoot, dir));
    let ignore = [this.options.cacheDir, ...vcsDirs];
    return {
      ignore
    };
  }

  getCacheKeys() {
    let assetGraphKey = (0, _utils.md5FromString)(`${this.cacheKey}:assetGraph`);
    let requestGraphKey = (0, _utils.md5FromString)(`${this.cacheKey}:requestGraph`);
    let snapshotKey = (0, _utils.md5FromString)(`${this.cacheKey}:snapshot`);
    return {
      assetGraphKey,
      requestGraphKey,
      snapshotKey
    };
  }

  async readFromCache() {
    if (this.options.disableCache) {
      return null;
    }

    let {
      assetGraphKey,
      requestGraphKey,
      snapshotKey
    } = this.getCacheKeys();
    let assetGraph = await this.options.cache.get(assetGraphKey);
    let requestGraph = await this.options.cache.get(requestGraphKey);

    if (assetGraph && requestGraph) {
      this.assetGraph = assetGraph;
      this.requestGraph = requestGraph;
      let opts = this.getWatcherOptions();

      let snapshotPath = this.options.cache._getCachePath(snapshotKey, '.txt');

      return this.options.inputFS.getEventsSince(this.options.projectRoot, snapshotPath, opts);
    }

    return null;
  }

  async writeToCache() {
    if (this.options.disableCache) {
      return;
    }

    let {
      assetGraphKey,
      requestGraphKey,
      snapshotKey
    } = this.getCacheKeys();
    await this.options.cache.set(assetGraphKey, this.assetGraph);
    await this.options.cache.set(requestGraphKey, this.requestGraph);
    let opts = this.getWatcherOptions();

    let snapshotPath = this.options.cache._getCachePath(snapshotKey, '.txt');

    await this.options.inputFS.writeSnapshot(this.options.projectRoot, snapshotPath, opts);
  }

}

exports.default = AssetGraphBuilder;
