// @parcel/core/lib/RequestTracker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.RequestGraph = void 0;

var _assert = _interopRequireDefault(require("assert"));

var _nullthrows = _interopRequireDefault(require("nullthrows"));

var _utils = require("@parcel/utils");

var _Graph = _interopRequireDefault(require("./Graph"));

var _utils2 = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const nodeFromFilePath = filePath => ({
  id: filePath,
  type: 'file',
  value: {
    filePath
  }
});

const nodeFromGlob = glob => ({
  id: glob,
  type: 'glob',
  value: glob
});

const nodeFromRequest = request => ({
  id: request.id,
  type: 'request',
  value: request
});

class RequestGraph extends _Graph.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "invalidNodeIds", new Set());

    _defineProperty(this, "incompleteNodeIds", new Set());

    _defineProperty(this, "globNodeIds", new Set());

    _defineProperty(this, "unpredicatableNodeIds", new Set());
  }

  // $FlowFixMe
  static deserialize(opts) {
    // $FlowFixMe Added in Flow 0.121.0 upgrade in #4381
    let deserialized = new RequestGraph(opts);
    deserialized.invalidNodeIds = opts.invalidNodeIds;
    deserialized.incompleteNodeIds = opts.incompleteNodeIds;
    deserialized.globNodeIds = opts.globNodeIds;
    deserialized.unpredicatableNodeIds = opts.unpredicatableNodeIds; // $FlowFixMe Added in Flow 0.121.0 upgrade in #4381 (Windows only)

    return deserialized;
  } // $FlowFixMe


  serialize() {
    return { ...super.serialize(),
      invalidNodeIds: this.invalidNodeIds,
      incompleteNodeIds: this.incompleteNodeIds,
      globNodeIds: this.globNodeIds,
      unpredicatableNodeIds: this.unpredicatableNodeIds
    };
  }

  addNode(node) {
    if (!this.hasNode(node.id)) {
      if (node.type === 'glob') {
        this.globNodeIds.add(node.id);
      }
    }

    return super.addNode(node);
  }

  removeNode(node) {
    this.invalidNodeIds.delete(node.id);
    this.incompleteNodeIds.delete(node.id);

    if (node.type === 'glob') {
      this.globNodeIds.delete(node.id);
    }

    return super.removeNode(node);
  }

  getRequestNode(id) {
    let node = (0, _nullthrows.default)(this.getNode(id));
    (0, _assert.default)(node.type === 'request');
    return node;
  }

  completeRequest(request) {
    this.invalidNodeIds.delete(request.id);
    this.incompleteNodeIds.delete(request.id);
  }

  replaceSubrequests(requestId, subrequestNodes) {
    let requestNode = this.getRequestNode(requestId);

    if (!this.hasNode(requestId)) {
      this.addNode(requestNode);
    }

    this.replaceNodesConnectedTo(requestNode, subrequestNodes, null, 'subrequest');
  }

  invalidateNode(node) {
    (0, _assert.default)(node.type === 'request');

    if (this.hasNode(node.id)) {
      this.invalidNodeIds.add(node.id);
      this.clearInvalidations(node);
      let parentNodes = this.getNodesConnectedTo(node, 'subrequest');

      for (let parentNode of parentNodes) {
        this.invalidateNode(parentNode);
      }
    }
  }

  invalidateUnpredictableNodes() {
    for (let nodeId of this.unpredicatableNodeIds) {
      let node = (0, _nullthrows.default)(this.getNode(nodeId));
      (0, _assert.default)(node.type !== 'file' && node.type !== 'glob');
      this.invalidateNode(node);
    }
  }

  invalidateOnFileUpdate(requestId, filePath) {
    let requestNode = this.getRequestNode(requestId);
    let fileNode = nodeFromFilePath(filePath);

    if (!this.hasNode(fileNode.id)) {
      this.addNode(fileNode);
    }

    if (!this.hasEdge(requestNode.id, fileNode.id, 'invalidated_by_update')) {
      this.addEdge(requestNode.id, fileNode.id, 'invalidated_by_update');
    }
  }

  invalidateOnFileDelete(requestId, filePath) {
    let requestNode = this.getRequestNode(requestId);
    let fileNode = nodeFromFilePath(filePath);

    if (!this.hasNode(fileNode.id)) {
      this.addNode(fileNode);
    }

    if (!this.hasEdge(requestNode.id, fileNode.id, 'invalidated_by_delete')) {
      this.addEdge(requestNode.id, fileNode.id, 'invalidated_by_delete');
    }
  }

  invalidateOnFileCreate(requestId, glob) {
    let requestNode = this.getRequestNode(requestId);
    let globNode = nodeFromGlob(glob);

    if (!this.hasNode(globNode.id)) {
      this.addNode(globNode);
    }

    if (!this.hasEdge(requestNode.id, globNode.id, 'invalidated_by_create')) {
      this.addEdge(requestNode.id, globNode.id, 'invalidated_by_create');
    }
  }

  invalidateOnStartup(requestId) {
    let requestNode = this.getRequestNode(requestId);
    this.unpredicatableNodeIds.add(requestNode.id);
  }

  clearInvalidations(node) {
    this.unpredicatableNodeIds.delete(node.id);
    this.replaceNodesConnectedTo(node, [], null, 'invalidated_by_update');
    this.replaceNodesConnectedTo(node, [], null, 'invalidated_by_delete');
    this.replaceNodesConnectedTo(node, [], null, 'invalidated_by_create');
  }

  respondToFSEvents(events) {
    for (let {
      path,
      type
    } of events) {
      let node = this.getNode(path); // sometimes mac os reports update events as create events
      // if it was a create event, but the file already exists in the graph,
      // then we can assume it was actually an update event

      if (node && (type === 'create' || type === 'update')) {
        for (let connectedNode of this.getNodesConnectedTo(node, 'invalidated_by_update')) {
          this.invalidateNode(connectedNode);
        }
      } else if (type === 'create') {
        for (let id of this.globNodeIds) {
          let globNode = this.getNode(id);
          (0, _assert.default)(globNode && globNode.type === 'glob');

          if ((0, _utils.isGlobMatch)(path, globNode.value)) {
            let connectedNodes = this.getNodesConnectedTo(globNode, 'invalidated_by_create');

            for (let connectedNode of connectedNodes) {
              this.invalidateNode(connectedNode);
            }
          }
        }
      } else if (node && type === 'delete') {
        for (let connectedNode of this.getNodesConnectedTo(node, 'invalidated_by_delete')) {
          this.invalidateNode(connectedNode);
        }
      }
    }

    return this.invalidNodeIds.size > 0;
  }

}

exports.RequestGraph = RequestGraph;

class RequestTracker {
  constructor({
    graph,
    farm,
    options
  }) {
    _defineProperty(this, "graph", void 0);

    _defineProperty(this, "farm", void 0);

    _defineProperty(this, "options", void 0);

    _defineProperty(this, "signal", void 0);

    this.graph = graph || new RequestGraph();
    this.farm = farm;
    this.options = options;
  } // TODO: refactor (abortcontroller should be created by RequestTracker)


  setSignal(signal) {
    this.signal = signal;
  }

  startRequest(request) {
    if (!this.graph.hasNode(request.id)) {
      let node = nodeFromRequest(request);
      this.graph.addNode(node);
    }

    this.graph.incompleteNodeIds.add(request.id);
    this.graph.invalidNodeIds.delete(request.id);
  }

  removeRequest(id) {
    this.graph.removeById(id);
  }

  storeResult(id, result) {
    let node = this.graph.getNode(id);

    if (node && node.type === 'request') {
      node.value.result = result;
    }
  }

  hasValidResult(id) {
    return this.graph.nodes.has(id) && !this.graph.invalidNodeIds.has(id) && !this.graph.incompleteNodeIds.has(id);
  }

  getRequestResult(id) {
    let node = (0, _nullthrows.default)(this.graph.getNode(id));
    (0, _assert.default)(node.type === 'request'); // $FlowFixMe

    let result = node.value.result;
    return result;
  }

  completeRequest(id) {
    this.graph.invalidNodeIds.delete(id);
    this.graph.incompleteNodeIds.delete(id);
  }

  rejectRequest(id) {
    this.graph.incompleteNodeIds.delete(id);

    if (this.graph.hasNode(id)) {
      this.graph.invalidNodeIds.add(id);
    }
  }

  respondToFSEvents(events) {
    return this.graph.respondToFSEvents(events);
  }

  hasInvalidRequests() {
    return this.graph.invalidNodeIds.size > 0;
  }

  getInvalidRequests() {
    let invalidRequests = [];

    for (let id of this.graph.invalidNodeIds) {
      let node = (0, _nullthrows.default)(this.graph.getNode(id));
      (0, _assert.default)(node.type === 'request');
      invalidRequests.push(node.value);
    }

    return invalidRequests;
  }

  replaceSubrequests(requestId, subrequestNodes) {
    this.graph.replaceSubrequests(requestId, subrequestNodes);
  }

  async runRequest(request) {
    let id = request.id;

    if (this.hasValidResult(id)) {
      return this.getRequestResult(id);
    }

    let {
      api,
      subRequests
    } = this.createAPI(id);

    try {
      this.startRequest({
        id,
        type: request.type,
        input: request.input
      });
      let result = await request.run({
        input: request.input,
        api,
        farm: this.farm,
        options: this.options
      });
      (0, _utils2.assertSignalNotAborted)(this.signal);
      this.completeRequest(id);
      return result;
    } catch (err) {
      this.rejectRequest(id);
      throw err;
    } finally {
      this.graph.replaceSubrequests(id, [...subRequests].map(subRequestId => (0, _nullthrows.default)(this.graph.getNode(subRequestId))));
    }
  }

  createAPI(requestId) {
    let subRequests = new Set();
    return {
      api: {
        invalidateOnFileCreate: glob => this.graph.invalidateOnFileCreate(requestId, glob),
        invalidateOnFileDelete: filePath => this.graph.invalidateOnFileDelete(requestId, filePath),
        invalidateOnFileUpdate: filePath => this.graph.invalidateOnFileUpdate(requestId, filePath),
        invalidateOnStartup: () => this.graph.invalidateOnStartup(requestId),
        storeResult: result => {
          this.storeResult(requestId, result);
        },
        runRequest: subRequest => {
          subRequests.add(subRequest.id);
          return this.runRequest(subRequest);
        }
      },
      subRequests
    };
  }

}

exports.default = RequestTracker;
