// @parcel/core/lib/worker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.runTransform = runTransform;
exports.runValidate = runValidate;
exports.runPackage = runPackage;

var _assert = _interopRequireDefault(require("assert"));

var _nullthrows = _interopRequireDefault(require("nullthrows"));

var _BundleGraph = _interopRequireDefault(require("./BundleGraph"));

var _Transformation = _interopRequireDefault(require("./Transformation"));

var _ReporterRunner = require("./ReporterRunner");

var _PackagerRunner = _interopRequireDefault(require("./PackagerRunner"));

var _Validation = _interopRequireDefault(require("./Validation"));

var _ParcelConfig = _interopRequireDefault(require("./ParcelConfig"));

var _utils = require("./utils");

require("@parcel/cache");

require("@parcel/package-manager");

require("@parcel/fs");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// register with serializer
(0, _utils.registerCoreWithSerializer)(); // Remove the workerApi type from the TransformationOpts and ValidationOpts types:
// https://github.com/facebook/flow/issues/2835

// TODO: this should eventually be replaced by an in memory cache layer
let parcelConfigCache = new Map();

function loadOptions(ref, workerApi) {
  return (0, _nullthrows.default)(workerApi.getSharedReference(ref // $FlowFixMe
  ));
}

async function loadConfig(cachePath, options) {
  var _parcelConfigCache$ge;

  let processedConfig = (_parcelConfigCache$ge = parcelConfigCache.get(cachePath)) !== null && _parcelConfigCache$ge !== void 0 ? _parcelConfigCache$ge : (0, _nullthrows.default)((await options.cache.get(cachePath)));
  let config = new _ParcelConfig.default( // $FlowFixMe
  processedConfig, options.packageManager, options.autoinstall);
  parcelConfigCache.set(cachePath, config);
  return config;
}

async function runTransform(workerApi, opts) {
  let {
    optionsRef,
    configCachePath,
    ...rest
  } = opts;
  let options = loadOptions(optionsRef, workerApi);
  let config = await loadConfig(configCachePath, options);
  return new _Transformation.default({
    workerApi,
    report: _ReporterRunner.reportWorker.bind(null, workerApi),
    options,
    config,
    ...rest
  }).run();
}

async function runValidate(workerApi, opts) {
  let {
    optionsRef,
    configCachePath,
    ...rest
  } = opts;
  let options = loadOptions(optionsRef, workerApi);
  let config = await loadConfig(configCachePath, options);
  return new _Validation.default({
    workerApi,
    report: _ReporterRunner.reportWorker.bind(null, workerApi),
    options,
    config,
    ...rest
  }).run();
}

function runPackage(workerApi, {
  bundle,
  bundleGraphReference,
  configRef,
  cacheKeys,
  optionsRef
}) {
  let bundleGraph = workerApi.getSharedReference(bundleGraphReference);
  (0, _assert.default)(bundleGraph instanceof _BundleGraph.default);
  let options = loadOptions(optionsRef, workerApi);
  let processedConfig = workerApi.getSharedReference(configRef // $FlowFixMe
  );
  let config = new _ParcelConfig.default(processedConfig, options.packageManager, options.autoinstall);
  return new _PackagerRunner.default({
    config,
    options,
    report: _ReporterRunner.reportWorker.bind(null, workerApi)
  }).getBundleInfo(bundle, bundleGraph, cacheKeys);
}
